src/server/pps/server/pipeline_controller.go

```diff
@@ -748,19 +796,20 @@ func (step *pcStep) getRC(ctx context.Context, expectation rcExpectation) (retEr
 // updated is already available to the caller in pc.rc, but update() may be
 // called muliple times if the k8s write fails. It may be helpful to think of
 // the rc passed to update() as mutable, while pc.rc is immutable.
-func (step *pcStep) updateRC(ctx context.Context, update func(rc *v1.ReplicationController)) error {
+func (step *pcStep) updateRC(ctx context.Context, update func(rc *v1.ReplicationController) bool) error {
+	rc := step.pc.env.KubeClient.CoreV1().ReplicationControllers(step.pc.namespace)
 
 	step.pc.pcMgr.limiter.Acquire()
 	defer step.pc.pcMgr.limiter.Release()
 
-	rc := step.pc.env.KubeClient.CoreV1().ReplicationControllers(step.pc.namespace)
-
-	newRC := *step.rc
 	// Apply op's update to rc
-	update(&newRC)
-	// write updated RC to k8s
-	if _, err := rc.Update(ctx, &newRC, metav1.UpdateOptions{}); err != nil {
-		return newRetriableError(err, "error updating RC")
+	newRC := *step.rc
+	if update(&newRC) {
```

Comment:
This is the main change that eliminates no-op scaling updates

Reply:
Is step a protocol buffer?  Is so, we should use proto.Clone instead of dereference then indirect.

Reply:
It's not a proto, but by investigating I found out that it does have a `DeepCopy()` method, which this now uses.